SYCL/ESIMD/api/functional/ctors/ctor_converting.hpp

```diff
@@ -0,0 +1,164 @@
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "../value_conv.hpp"
+#include "common.hpp"
+
+namespace esimd = sycl::ext::intel::experimental::esimd;
+
+namespace esimd_test::api::functional::ctors {
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct initializer {
+  static std::string get_description() { return "initializer"; }
+
+  template <typename SrcT, typename DstT, int NumElems>
+  static void call_simd_ctor(const SrcT *const ref_data, DstT *const out) {
+    esimd::simd<SrcT, NumElems> input_simd;
+    input_simd.copy_from(ref_data);
+
+    esimd::simd<DstT, NumElems> output_simd = input_simd;
+    output_simd.copy_to(out);
+  }
+};
+
+template <typename T, int NumElems, typename ContextT>
+class ConvCtorTestDescription : public ITestDescription {
+public:
+  ConvCtorTestDescription(size_t index, T retrieved_val, T expected_val,
+                          const std::string &src_data_type,
+                          const std::string &dst_data_type)
+      : m_src_data_type(src_data_type), m_dst_data_type(dst_data_type),
+        m_retrieved_val(retrieved_val), m_expected_val(expected_val),
+        m_index(index) {}
+
+  std::string to_string() const override {
+    // TODO: Make strings for fp values more short during failure output, may be
+    // by using hex representation
+    std::string log_msg("Failed for converting from simd<");
+
+    log_msg += m_src_data_type + ", " + std::to_string(NumElems) + ">";
+    log_msg +=
+        ", to simd<" + m_dst_data_type + ", " + std::to_string(NumElems) + ">";
+    log_msg += ", with context: " + ContextT::get_description();
+    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
+    log_msg += ", expected: " + std::to_string(m_expected_val);
+    log_msg += ", at index: " + std::to_string(m_index);
+
+    return log_msg;
+  }
+
+private:
+  const std::string m_src_data_type;
+  const std::string m_dst_data_type;
+  const T m_retrieved_val;
+  const T m_expected_val;
+  const size_t m_index;
+};
+
+template <typename, int, typename> class Kernel;
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename TestCaseT, typename SrcT, typename DimT, typename DstT>
+class run_test {
+  static constexpr int NumElems = DimT::value;
+
+public:
+  bool operator()(sycl::queue &queue, const std::string &src_data_type,
+                  const std::string &dst_data_type) {
+    bool passed = true;
+    const std::vector<SrcT> ref_data =
+        generate_ref_conv_data<SrcT, DstT, NumElems>();
+
+    // If current number of elements is equal to one, then run test with each
+    // one value from reference data.
+    // If current number of elements is greater than one, then run tests with
+    // whole reference data.
+    if constexpr (NumElems == 1) {
```

Comment:
Minor: have you verified compilation/run for this condition? Looks like it's always false with current test coverage. It's great to have a possibility to switch support for `simd<T,1>` by single line in tests though.

What would you say about using the `ESIMD_TESTS_FULL_COVERAGE` in cpp files? So we could verify test works for `simd<T,1>` at any time by simple macro definition.

Reply:
I have verified this and we don't have any syntaxes error in this branch, also I don't wont to add a macros to enable this branch and I don't won't to delete this branch, it may be useful in the future